File: Python3/733. Flood Fill.py
class Solution:
    def floodFill(self, image, sr, sc, newColor):
        """
        :type image: List[List[int]]
        :type sr: int
        :type sc: int
        :type newColor: int
        :rtype: List[List[int]]
        """
        if image[sr][sc] == newColor: return image # 72ms
        originColor = image[sr][sc]
        rlen = len(image)
        clen = len(image[0])
        def fillnewColor(r, c):
            if image[r][c] == originColor:
                image[r][c] = newColor
                if r > 0: fillnewColor(r-1, c)
                if r < rlen - 1: fillnewColor(r+1, c)
                if c > 0: fillnewColor(r, c-1)
                if c < clen - 1: fillnewColor(r, c+1)
        fillnewColor(sr, sc)
        return image

File: Golang/733. Flood Fill.go
func floodFill(image [][]int, sr int, sc int, newColor int) [][]int { // 28ms
    originColor := image[sr][sc]
    if originColor == newColor {
        return image
    }
    fillnewColor(&image, sr, sc, originColor, newColor)
    return image   
}

func fillnewColor(image *[][]int, r, c, originColor, newColor int) {
    if (*image)[r][c] == originColor {
        (*image)[r][c] = newColor
        if r > 0 {
            fillnewColor(image, r-1, c, originColor, newColor)
        }
        if r < len(*image) - 1 {
            fillnewColor(image, r+1, c, originColor, newColor)
        }
        if c > 0 {
            fillnewColor(image, r, c-1, originColor, newColor)
        }
        if c < len((*image)[0]) - 1 {
            fillnewColor(image, r, c+1, originColor, newColor)
        }
    }
}
